clang/runtime/dpct-rt/include/dpct/device.hpp

```diff
@@ -778,7 +812,7 @@ class dev_mgr {
   /// for the current thread.
   const unsigned int DEFAULT_DEVICE_ID = 0;
   /// thread-id to device-id map.
-  std::map<unsigned int, unsigned int> _thread2dev_map;
+  std::map<unsigned int, std::stack<unsigned int>> _thread2dev_map;
```

Comment:
```suggestion
  static thread_local std::stack<unsigned> _curr_id;
```

Reply:
Can we use `thread_local` variable to replace the map? And then we don't need to use mutex in `select_device` `push_device`.

Reply:
Using two thread scope variables is a cleaner idea.

But it seems like mutex lock is also supposed to prevent race condition for vector `_devs`. I came to this conclusion because it's used in `get_device` where there is no access of `thread2dev_map`. Note: `_devs` is written in ctor `dev_mgr()` and in method `filter` and read in multiple places.

Does it make sense to also make the `_devs` thread_local? IMO no.

Reply:
I think `static thread_local unsigned _curr_id = 0;` can work for the initialization.

I don't mean that all the mutex usage will be removed. Just remove the used for avoiding race condition of `thread2dev_map`

Reply:
`_dev_stack` will be a `thread_local` and we will use mutex for access to `_devs`.